torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -232,6 +233,35 @@ class OptimStateKeyType(Enum):
     PARAM_NAME = auto()
     PARAM_ID = auto()
 
+class _ExecOrderData():
+    """
+    This contains the data used for validating execution order across ranks.
+
+    Attributes:
+        param_to_param_index (Dict[FlatParameter, int]): Mapping from FSDP
+            flattened parameter to a unique parameter index consistent across
+            ranks.
+        is_first_iter (bool): Whether executing in the first iteration or not.
+        param_order (List[int]): Order that parameters participate in the
+            forward pass; constructed on the first iteration and validated
+            against in subsequent iterations.
+        index (int): Index tracking the position in ``param_order``
+            when validating the forward pass execution order in subsequent
+            iterations.
+    """
+    param_to_param_index: Dict[FlatParameter, int] = {}
+    # Modified in the first iteration:
+    is_first_iter: bool = True
+    param_order: List[int] = []
+    # Modified in the subsequent iterations:
+    index: int = 0
+
+    def reset(self):
+        """Called in :meth:`_wait_for_post_backward` to reset data for the next
+        iteration."""
+        self.is_first_iter = False
+        self.index = 0
```

Comment:
It looks like `param_order` also gets modified, do we not need to reset this?

Reply:
Yes, `param_to_param_index` is initialized once and never changed, and param_order` is constructed over the course of the first iteration and not changed thereafter.